### PR TITLE
Add SegWit support

### PR DESCRIPTION
**Pull request by jameshilliard**
This is an initial attempt at adding SegWit support, it appears to mine valid blocks on testnet.
I mined these two blocks which contain SegWit transactions:
https://www.blocktrail.com/tBTC/block/0000000000001117c46477e777a76c77857187d97aca5d0f621b29ed7dd45625
https://www.blocktrail.com/tBTC/block/00000000000017db788821b7e36f89c670b14888c3a946d8efdc72963115d755

### CODE REVIEW

**Comment by ahmedbodi** [comment]
NACK! This is going to break POS coin support. There needs to be a config option which specifies something like segWitEnabled: true. if so then use this code, otherwise the old stuff

**Comment by jameshilliard** [comment]
@ahmedbodi how would this affect POS coins? the witness commitment does not get inserted unless the node specifically requests it in the GBT response

**Comment by ahmedbodi** [comment]
the GBT call, Old POS coins (even new ones really) expect the gbt call to pass nothing as params. This passes the whole rules and other stuff

File: lib/transactions.js [hunk 1]
```diff
@@ -172,6 +172,14 @@ var generateOutputTransactions = function(poolRecipient, recipients, rpcData){
         poolRecipient
     ]));
 
+    if (rpcData.default_witness_commitment !== undefined){
```
Comment (jameshilliard) [inline lib/transactions.js:207]: If there is no "default_witness_commitment" in the GBT response the witness commitment is not inserted.

File: lib/pool.js [hunk 2]
```diff
@@ -144,23 +144,25 @@ var pool = module.exports = function pool(options, authorizeFn){
...
-                });
-                if (synced){
-                    syncedCallback();
-                }
-                else{
-                    if (displayNotSynced) displayNotSynced();
-                    setTimeout(checkSynced, 5000);
+            _this.daemon.cmd('getblocktemplate',
+                [{"capabilities": [ "coinbasetxn", "workid", "coinbase/append" ], "rules": [ "segwit" ]}],
+                function(results){
```
Comment (ahmedbodi) [inline lib/pool.js]: Here, this would break POS coins. they expect to get [] not any content

**Comment by jameshilliard** [comment]
@ahmedbodi what about "capabilities"?

File: lib/pool.js [hunk 3]
```diff
@@ -694,7 +696,7 @@ var pool = module.exports = function pool(options, authorizeFn){
 
     function GetBlockTemplate(callback, force){
         _this.daemon.cmd('getblocktemplate',
-            [{"capabilities": [ "coinbasetxn", "workid", "coinbase/append" ]}],
```
Comment (jameshilliard) [inline lib/pool.js]: Wouldn't this cause an issue for POS coins as well then?

**Review by ahmedbodi — CHANGES_REQUESTED** [review]

File: lib/pool.js (hunk 3 above)
Comment (ahmedbodi) [reply lib/pool.js]: weird. it certainly should do, i wonder when it was last changed

File: lib/pool.js (hunk 2 above)
Comment (jameshilliard) [reply lib/pool.js]: It should now only send the params for POW coins

**Review by ahmedbodi — APPROVED** [review]
Untested. But Looks OK

**Comment by btcdrak** [comment]
@jameshilliard Please rebase.

This patch is working on mainnet, the following blocks were mined using this PR

http://explorer.litecoin.net/block/92f1efabbac74a0362a900aaf85b9419b8ef7daa2f7a347d5b917caf402a5ac7
http://explorer.litecoin.net/block/e2ce3731d61cc4c918fb76a1f46755a3ea05298e93f69b80ac96b85e03be04a4

**Comment by jameshilliard** [comment]
rebased

**Comment by Infernoman** [comment]
Going to make a slight change to this, cleaning up the second getblocktemplate for peercoin, and segwit. keep an eye out on the next commit.

**Comment by Infernoman** [comment]
https://github.com/UNOMP/node-merged-pool/commit/c79e829d151fc90dfb2cff0399b995d76e2ce036

**Comment by jstefanop** [comment]
@Infernoman Just to confirm, this patch will signal segwit blocks if I mine them? Thinking about setting up a pool to help Litecoin activate segwit. Any issues with merge mining segwit blocks with other coins?

**Pull request merged**